README.MD
# BADGARIO - A 💩 Version of Agar.io
By Haoda Fan (http://haodafan.com)
Current version: 0.0.8D (playable)

## HOW TO INSTALL AND PLAY
__Note that you will need both Python and Pygame to play this game. If you have both, skip to setup.__
### How to install Python 3.6.3 if you don't have it:
Step 1: Download python 3.6.3 from https://www.python.org/downloads/release/python-363/
    (download the executable installer if you have Windows)

Step 2: Open the file and check "Add Python 3.6 to PATH"

Step 3: Click Install Now and install

Step 4: Open command prompt and type 'Python' to verify that you've installed it properly


### How to install pygame if you don't have it:
Step 1: open up command prompt or similar

Step 2: enter command "pip install pygame"


### How to setup game:
Step 1: Download the contents of this repository (click on the 'clone or download' button on the top right of the file navigation and click 'DOWNLOAD ZIP')

Step 2: Extract to any folder of your choosing

Step 3: Run play-badgario.py

Note that if you're more tech savvy, you can choose to git clone this repository instead. 

## INFORMATION ABOUT HOW I MADE THIS GAME AND STUFF
This game was inspired by agar.io and much of the code was based off of a pygame tutorial game known as squirrel eat squirrel, which has similar mechanics. You can see it here:  https://inventwithpython.com/makinggames.pdf (page 200).

Resources used:
- Python 3.6.3
- PyGame
- MS Paint (for the default boss lel)
- My blood, sweat, and tears. Just kidding, only tears.

## Special Thanks:
### Voice actors for default bosses:

* /tanner - **Tanner Van Teeffelen**  (Listen to his music at https://soundcloud.com/auditory-ascendance)

### Other people I would like to thank: 

* Jackie Yuan Xu for initially helping me with pygame and inspiring me to make this game :> (check out his github at https://github.com/OnePunMan)


changelog.txt
#################################
October 17th 2017
Version: 0.0.1

Game is playable, there are win and lose conditions.
Camera follows the player around
There are always a total of 30 balls floating around the active area around your camera
Balls are deleted if they go too far out, new balls are generated based on the size of your ball
You can eat other balls, other balls can eat you

#################################
October 18th 2017
Version: 0.0.2

A simple boss is added in.
The boss:
- follows you around
- is slightly slower, but once it's outside the active area, a catchup mechanic prevents him from getting too far from the player

#################################
October 18th 2017
Version: 0.0.3

Boss feature additions:
- Boss will run away from you after you reach a certain size
- Engulfing mechanics with Boss created
- Boss will detect when it comes into camera view, and when it touches you
NEW VICTORY CONDITION: You win if you eat the boss!

#################################
October 18th 2017
Version: 0.0.4
Added sound:
- Boss will play sounds when it detects coming into camera view, touching player, and the turnaround
- Boss will play sounds when it eats and is eaten

Minor changes:
- Changed img.png to Tanner


#################################
October 19th 2017
Version: 0.0.4B
Fixed bug where boss would never go diagonally down-left


#################################
October 20th 2017
Version: 0.0.5

New Feature:
- Added an arrow that points in the direction of the boss when he is out of screen

Bug fixes:
- Fixed the boss's moving AI bug where he would not move down unless you are directly below him
- Boss's ai works as intended now.

Changed so that the boss stops playing sounds after the turnaround

#################################
October 20th 2017
Version 0.0.5B

Removed the build folder - it wasn't working properly anyway
Also deleted the pycache folder

#################################
October 23rd 2017
Version 0.0.6

New Feature: Main Menu (play_badgario.py)
Added a main menu with three options:
1. Play game - launches the main game
2. Game settings - launches a text file with instructions on how to manually change game settings
3. Customize Boss - launches a text file with instructions on how to customize the ingame boss

Also added instructions on the latter two things

#################################
October 24th 2017
Version 0.0.7

New Feature: Pause
When a player presses 'p', the game will pause. All movement is frozen, and the background will turn grey. You press 'p' again to unpause.

#################################
November 14th 2017
Version 0.0.8

New Feature: Dynamic Player Speed AND New loss condition: You are too fat!
As the player gets bigger, he will also become slower. This is done through a series of stages where when the player passes each size stage, the player loses 1 movespeed.
If the player's movespeed becomes the same or lower than the boss, then he loses because he would never be able to catch up.

New Feature: Return to Menu
When the player hits 'escape' or 'm', it will return to the main menu screen.
(note for developers: I have changed the runGame function so that it will return a boolean indicating whether to RETURN TO MENU (True) or RESTART (False))

Changes:
- All victory/defeat text is gray
- Set it so that the player will be displayed IN FRONT OF the boss when he is larger.

#################################
November 14th 2017
Version 0.0.8B

Cheats added:
- Press control+'=' to increase your size value by 5.
- Press control+'-' to decrease your size value by 5. (NOTE THAT DECREASING YOUR SIZE BY CHEATS DOES NOT MAKE YOU FASTER)

Updated README.MD to include special thanks section for my voice actor(s)

#################################
November 15th 2017
Version 0.0.8C

Changes:
- You can now press 'r' to restart the game no matter if you win, lose, or become too fat.
- Gave the press 'r' to restart text screen to all end-game screens.

Bug fixes:
Fixed bug where the Turnaround flag will not happen if you use cheats to increase size.
Fixed bug where the game will not end after you become too fat, due to the endgame statement being nested in the dynamic speed if statement.

################################
November 17th 2017
Version 0.0.8D

Changes:
Added a new color: Turqoise! :D
The default game is a bit too easy, so I made some significant balance changes:
 - Changed the minimum engulfing size change from +2 to +1
 - Changed the boss size from 200 to 210
 - Changed the staging system
	- Starting speed changed from 8 to 13
	- Stage separation (The amount of size you must gain to lose speed) changed from 50 to 25
	- You will become TOO FAT at size 250 instead of size 275

Personally, I'm not sure if this makes the game better or worse. I'll get my playtesters to try it out... IF I HAD ANY. :'(

#################################
November 28th 2017
Version 0.0.8E

Changes:
Balance of the game changed again. The ball is initially too fast, and becomes too fat too early.
The new default game is somewhere between the settings in 0.0.8D and the original 0.0.8
You can view stats in the original settings in the source code for badgario.py and config.py
Here are the changes:
 - Changed the boss size from 210 to 215
 - Changed the smallest possible enemy ball size from 8 to 10
 - Changed the staging system
  - Changed the default start size from 25 to 33
  - Changed the stage separation from 25 to 33
  - You will become TOO FAT at size 264 instead of size 250



STREEEETCH GOAL
#################################
Version 1.0.0
- Game will be fully playable
- Boss sounds will play when it detects things
- Can customize boss using addon program or simple modding
- Game is fully configurable and customizable using addon program


play_badgario.py
# This Menu is in Playable condition! PLAYABLE!! 

# ----------------------------------------------------
from badgario import *
from config import *

#import webbrowser #To externally open files
import os
# ----------------------------------------------------

#Initialize pygame
pygame.init()

class GameMenu():
    def __init__(self, screen):
        self.screen = screen
        self.bgColor = BLACK
        self.clock = pygame.time.Clock()

        #More of Initializing pygame
        pygame.display.set_caption('Shitty Agario')
        self.SURF = screen

        self.SMALLESTFONT = pygame.font.Font('freesansbold.ttf', 14)
        self.SMALLERFONT = pygame.font.Font('freesansbold.ttf', 20)
        self.BASICFONT = pygame.font.Font('freesansbold.ttf', 32)
        self.BIGFONT = pygame.font.Font('freesansbold.ttf', 72)


    ### Main running menu function ###
    def run(self):
        mainLoop = True

        #Some button boolean flags
        playHover = False;
        configHover = False;
        bossHover = False;
        ####################
        ## Main menu loop ##
        ####################
        while mainLoop:
            #Limit frame speed to FPS
            self.clock.tick(15)


            ## Screen ##

            #Drawing the background
            self.SURF.fill(WHITE)
            #pygame.display.flip() #To be honest, not sure what this is supposed to do.


            #Titles
            textTitleSurf = self.BIGFONT.render("Badgario - A shitty version of Agar.io", True, BLACK)
            textTitleRect = textTitleSurf.get_rect()
            textTitleRect.center = (HALF_SCREENSIZE_X, HALF_SCREENSIZE_Y - CAMERASLACK - 72)

            textSubSurf = self.SMALLERFONT.render("A silly half-baked game by Haoda Fan - the master of silly half-baked things.", True, BLACK)
            textSubRect = textSubSurf.get_rect()
            textSubRect.center = (HALF_SCREENSIZE_X, HALF_SCREENSIZE_Y - CAMERASLACK)

            #Drawing the titles
            self.SURF.blit(textTitleSurf, textTitleRect)
            self.SURF.blit(textSubSurf, textSubRect)

            ## Buttons ##

            #Button 1: Play the game

            if playHover:
                textPlayColor = WHITE
                backPlayColor = BLACK
            else:
                textPlayColor = BLACK
                backPlayColor = WHITE

            pygame.draw.circle(self.SURF, backPlayColor, (HALF_SCREENSIZE_X, HALF_SCREENSIZE_Y + CAMERASLACK + 25), CAMERASLACK, 0)


            textPlaySurf = self.SMALLERFONT.render("Play game!", True, textPlayColor)
            textPlayRect = textPlaySurf.get_rect()
            textPlayRect.center = (HALF_SCREENSIZE_X, HALF_SCREENSIZE_Y + CAMERASLACK + 25)
            self.SURF.blit(textPlaySurf, textPlayRect)

            #Button 2: How to configure game

            if configHover:
                textConfigColor = WHITE
                backConfigColor = BLUE
            else:
                textConfigColor = BLUE
                backConfigColor = WHITE

            pygame.draw.circle(self.SURF, backConfigColor, (50 + CAMERASLACK, HALF_SCREENSIZE_Y + CAMERASLACK + 25), CAMERASLACK, 0)

            textConfigSurf = self.SMALLESTFONT.render("Game Settings", True, textConfigColor)
            textConfigRect = textConfigSurf.get_rect()
            textConfigRect.center = (50 + CAMERASLACK, HALF_SCREENSIZE_Y + CAMERASLACK + 25)
            self.SURF.blit(textConfigSurf, textConfigRect)

            #Button 3: How to customize boss

            if bossHover:
                textBossColor = WHITE
                backBossColor = RED
            else:
                textBossColor = RED
                backBossColor = WHITE

            pygame.draw.circle(self.SURF, backBossColor, (SCREENSIZE_X - CAMERASLACK - 50, HALF_SCREENSIZE_Y + CAMERASLACK + 25), CAMERASLACK, 0)

            textBossSurf = self.SMALLESTFONT.render("Customize Boss", True, textBossColor)
            textBossRect = textBossSurf.get_rect()
            textBossRect.center = (SCREENSIZE_X - CAMERASLACK - 50, HALF_SCREENSIZE_Y + CAMERASLACK + 25)
            self.SURF.blit(textBossSurf, textBossRect)

            #Coming soon! Buttons that do the work for you! (you lazy asshole)

            #Mouse recognition
            mouse = pygame.mouse.get_pos()
            #Note mouse[0] = x, mouse[1] = y

            # Mouse Hovering effects
            if GameMenu.isInRect(mouse[0], mouse[1], textPlayRect):
                playHover = True
                print("Hovering over play")
            elif GameMenu.isInRect(mouse[0], mouse[1], textConfigRect):
                configHover = True
                print("Hovering over config")
            elif GameMenu.isInRect(mouse[0], mouse[1], textBossRect):
                bossHover = True
                print("Hovering over boss")
            else:
                playHover = False
                configHover = False
                bossHover = False

            # Mouse clicking effects
            mouseClick = pygame.mouse.get_pressed()
            if mouseClick == (1,0,0) and playHover:
                time.sleep(0.2)
                main()
            elif mouseClick == (1,0,0) and configHover:
                os.system("notepad.exe more_instructions/config.txt")
            elif mouseClick == (1,0,0) and bossHover:
                os.system("notepad.exe more_instructions/customboss.txt")

            ## Event Listeners ##
            for event in pygame.event.get():
                # Quitting
                if event.type == pygame.QUIT:
                    mainLoop = False
                    terminate()

            #Updates the display
            pygame.display.update()

    # -------------------- #
    ### HELPER FUNCTIONS ###
    # -------------------- #
    def isInRect(x, y, objRect):
        newRect = pygame.Rect(x-1, y-1, 2, 2)
        return objRect.colliderect(newRect)


if __name__ == "__main__":
    #Create the screen
    screen = pygame.display.set_mode((SCREENSIZE_X, SCREENSIZE_Y), 0, 32)
    pygame.display.set_caption("Badgario - A shitty agario, by Haoda Fan")
    gm = GameMenu(screen)
    gm.run()


setup_badgario.py
# THIS PORTION IS UNUSED AND STILL IN DEVELOPMENT.
#

# ----------------------------------------------------
from badgario import *
# ----------------------------------------------------

pygame.init()

class GameMenu():
    def __init__(self, screen):
        self.screen = screen
        self.bgColor = BLACK
        self.clock = pygame.time.Clock()

        ## Main running menu function ##
        def run(self):
            mainLoop = True
            
            ## Main menu loop ##
            while mainLoop:
                #Limit frame speed to FPS
                self.clock.tick(FPS)

                ## Event Listeners ##
                for event in pygame.event.get():
                    if event.type == pygame.QUIT:
                        mainLoop = False
                        terminate()
                    

    if __name__ == "__main__":
        #Create the screen
        screen = pygame.display.set_mode((SCREENSIZE_X, SCREENSIZE_Y), 0, 32)
        pygame.display.set_caption("Badgario - A shitty agario, by Haoda Fan")
        gm = GameMenu(screen)
        gm.run()


move-move-i-gotta-go-to-the-bathroom.py
# November 22nd 2017
# Haoda Fan
# This module provides all the necessary functions to move the files required to customize bosses.

# ----------------------------------------------------------------
# Move! Outta my way!
# I gotta go to the bathroom!
# You gotta move! I gotta poop!
# I gotta poop I gotta poop I gotta poop I gotta poop!
# *trumpets*
# MOVE! My belly hurts so bad, and now I GOTTA CLIMB THE STAIRS?!
# OH SHIT I LET A LITTLE OUT
# *trumpets*
# HURRY!! GET OUT OF THE BATHROOM!!
# I gotta fucking shit so hard!
# MOVE! I GOTTA POOP! YOU REALLY NEED TO MOVE!
# GET OUT OF THE BATHROOM!!
# MOOOVE I GOTTA FUCKING POOP!!!!!!
# ----------------------------------------------------------------

#### LIBRARIES ####

import shutil
import os

#### DIRECTORIES ####
thisdir = os.getcwd() #current directory

ACTIVE_DIR = thisdir + "\\data\\active"
BOSSES_DIR = thisdir + "\\data\\bosses"

## More Variables ##
fileNames = ["\\img.png", "\\eat.ogg", "\\eaten.ogg", "\\ohno.ogg", "\\touch.ogg", "\\view.ogg"]

#### FUNCTIONS ####

# This function activates a particular boss
def activate(folderName):
    jobsDone = False
    
    #Loops through every filename
    for (fileName in fileNames):
        shutil.copy2(BOSSES_DIR + "\\" + folderName + fileName, ACTIVE_DIR + fileName)

    jobsDone = True
    return jobsDone
    
    
# This function returns the number of custom bosses
def bossCount():
    # Based on code from
    # https://arstechnica.com/civis/viewtopic.php?f=20&t=1111535
    numfiles = return len([f for f in os.listdir(BOSSES_DIR)) and f[0] != '.'])


config.py
# CONFIGURATION PYTHON FILE

# GENERAL SETTINGS
SCREENSIZE_X = 1366
SCREENSIZE_Y = 768
HALF_SCREENSIZE_X = int(SCREENSIZE_X/2) #(do not touch)
HALF_SCREENSIZE_Y = int(SCREENSIZE_Y/2) #(do not touch)

FPS = 30

CAMERASLACK = 90

# GAME SETTINGS
NAME = 'Player' #Player name
MOVERATE = 13 #Player initial speed

    # Set player speed and player speed decrease rate

    # 0.08 settings: Starting speed: 9, Starting size: 25, Rate: -1 SPEED / 50 SIZE
    #       SIZE    25  -  75  -  125  -  175  -  225  -  275
    #       SPEED    9  -   8  -   7   -   6   -   5   -   4

    # 0.08D settings: Starting speed: 13, Starting Size: 25, Rate: -1 SPEED / 25 SIZE
    #       SIZE    25  -  50  -  75  -  100  -  125  -  150  -  175  -  200  -  225  -  250
    #       SPEED   13  -  12  -  11  -   10  -   9   -   8   -   7   -   6   -   5   -   4

    # 0.08E settings: Starting speed: 11, Starting Size: 33, Rate: -1 SPEED / 33 SIZE
    #       SIZE    33  -  66  -  99  -  132  -  156  -  198  -  231  -  264
    #       SPEED   11  -  10  -   9  -   8   -   7   -   6   -   5   -   4
    # Current settings: 0.08E
STARTSIZE = 33
STAGESIZE = 33
EATSIZE = STARTSIZE - 15 # not implemented
WINSIZE = 200 # not implemented

GAMEOVERTIME = 4

NUMENEMIES = 30
ENEMYMINSIZE = 10
MINSPEED = 1
MAXSPEED = 4
DIRCHANGEFREQ = 3 # % chance of direction change per frame

# Not implemented
ENGULF_THRESHOLD_MULTIPLIER = 1.3 #You have to be this much times the size to engulf
ENGULF_THRESHOLD_CONSTANT = 20 #You have to be this many sizes bigger to engulf

ENGULF_EFFICIENCY = 0.5 #How much size you gain from engulfing the other object

#BOSS SETTINGS
BOSS_SIZE = 215
BOSS_SPEED = 4

BOSS_START_X = SCREENSIZE_X
BOSS_START_Y = SCREENSIZE_Y

#CUSTOMIZED BOSS (Do not touch)
BOSS_DIR = "data/active/img.png"
EAT_DIR = "data/active/eat.ogg"
EATEN_DIR = "data/active/eaten.ogg"
OHNO_DIR = "data/active/ohno.ogg"
VIEW_DIR = "data/active/view.ogg"
TOUCH_DIR = "data/active/touch.ogg"

#RED ARROW INDICATOR (Do not touch)
ARROW_PAD = 50
ARROW_SIZE = 40
REDUP_DIR = 'redUp.png'
REDUPLEFT_DIR = 'redUpRight.png'


# CONSTANTS (do not touch)
LEFT = 'left'
RIGHT = 'right'

# COLOUR CONSTANTS
WHITE  = (255, 255, 255)
BLACK  = (  0,   0,   0)
GRAY   = ( 64,  64,  64)

RED    = (255,   0,   0)
GREEN  = (  0, 255,   0)
BLUE   = (  0,   0, 255)
YELLOW = (255, 255,   0)
PURPLE = (128,   0, 128)
TURQ   = ( 50, 198, 166) #TURQOISE :D

BALLCOLOUR = [RED, GREEN, BLUE, YELLOW, PURPLE, TURQ]
